folding-schemes/src/folding/hypernova/decider_eth.rs

```diff
@@ -0,0 +1,309 @@
+        Proof = KZGProof<C1>,
+    >,
+    // enforce that the CS2 is Pedersen commitment scheme, since we're at Ethereum's EVM decider
+    CS2: CommitmentScheme<C2, ProverParams = PedersenParams<C2>>,
+    S: SNARK<C1::ScalarField>,
+    FS: FoldingScheme<C1, C2, FC>,
+    <C1 as CurveGroup>::BaseField: PrimeField,
+    <C2 as CurveGroup>::BaseField: PrimeField,
+    <C1 as Group>::ScalarField: Absorb,
+    <C2 as Group>::ScalarField: Absorb,
+    C1: CurveGroup<BaseField = C2::ScalarField, ScalarField = C2::BaseField>,
+    for<'b> &'b GC1: GroupOpsBounds<'b, C1, GC1>,
+    for<'b> &'b GC2: GroupOpsBounds<'b, C2, GC2>,
+    // constrain FS into HyperNova, since this is a Decider specifically for HyperNova
+    HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2, MU, NU, false>: From<FS>,
+    crate::folding::hypernova::ProverParams<C1, C2, CS1, CS2, false>:
+        From<<FS as FoldingScheme<C1, C2, FC>>::ProverParam>,
+    crate::folding::hypernova::VerifierParams<C1, C2, CS1, CS2, false>:
+        From<<FS as FoldingScheme<C1, C2, FC>>::VerifierParam>,
+{
+    type PreprocessorParam = (FS::ProverParam, FS::VerifierParam);
+    type ProverParam = (S::ProvingKey, CS1::ProverParams);
+    type Proof = Proof<C1, CS1, S>;
+    /// VerifierParam = (pp_hash, snark::vk, commitment_scheme::vk)
+    type VerifierParam = (C1::ScalarField, S::VerifyingKey, CS1::VerifierParams);
+    type PublicInput = Vec<C1::ScalarField>;
+    type CommittedInstance = ();
+
+    fn preprocess(
+        mut rng: impl RngCore + CryptoRng,
+        prep_param: &Self::PreprocessorParam,
+        fs: FS,
+    ) -> Result<(Self::ProverParam, Self::VerifierParam), Error> {
+        let circuit =
+            DeciderEthCircuit::<C1, GC1, C2, GC2, CS1, CS2>::from_hypernova::<FC, MU, NU>(
+                fs.into(),
+            )
+            .unwrap();
+
+        // get the Groth16 specific setup for the circuit
+        let (g16_pk, g16_vk) = S::circuit_specific_setup(circuit, &mut rng).unwrap();
+
+        // get the FoldingScheme prover & verifier params from HyperNova
+        #[allow(clippy::type_complexity)]
+        let hypernova_pp: <HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2, MU, NU,false> as FoldingScheme<
+            C1,
+            C2,
+            FC,
+        >>::ProverParam = prep_param.0.clone().into();
+        #[allow(clippy::type_complexity)]
+        let hypernova_vp: <HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2, MU, NU,false> as FoldingScheme<
+            C1,
+            C2,
+            FC,
+        >>::VerifierParam = prep_param.1.clone().into();
+        let pp_hash = hypernova_vp.pp_hash()?;
+
+        let pp = (g16_pk, hypernova_pp.cs_pp);
+        let vp = (pp_hash, g16_vk, hypernova_vp.cs_vp);
+        Ok((pp, vp))
+    }
+
+    fn prove(
+        mut rng: impl RngCore + CryptoRng,
+        pp: Self::ProverParam,
+        folding_scheme: FS,
+    ) -> Result<Self::Proof, Error> {
+        let (snark_pk, cs_pk): (S::ProvingKey, CS1::ProverParams) = pp;
```

Comment:
Notice that after the destructor, we just pass references to the internal functions called. Namely:
- `prove_with_challenge`
- `S::prove`.
So my suggestion is to just send a reference for `pp`.

Reply:
Not sure I see this one, we use `&snark_pk` and `&cs_pk`, but not `&pp` which contains the two first ones (without reference).

Reply:
But if you pass the tuple by value, you need to copy all it's items. SO you're copying all.
You can pass a touple with two references inside. But the change would be significantly bigger.

Reply:
but then the two references inside the tuple would need lifetime annotations, no?

Reply:
Ye... Which is unfortunate. Tuples in general are..
Let's leave it like this then.. It's sad. But for now I assume is better than loosing a lot of time with it

Reply:
(small joke: we would loose a lot of time, but we would gain a *lifetime* hehe)